src/countBy.test.ts

```diff
@@ -0,0 +1,112 @@
+import { countBy } from "./countBy";
+import { pipe } from "./pipe";
+
+describe("dataFirst", () => {
+  test("countBy", () => {
+    const data = [1, 2, 3, 2, 1, 5];
+    const result = countBy(data, (x) => x);
+
+    expect(result).toStrictEqual({ 1: 2, 2: 2, 3: 1, 5: 1 });
+  });
+
+  test("array of strings", () => {
+    const data = ["a", "b", "c", "B", "A", "a"];
+    const result = countBy(data, (x) => x.toLowerCase());
+
+    expect(result).toStrictEqual({ a: 3, b: 2, c: 1 });
+  });
+
+  test("array of objects", () => {
+    const data = [
+      { id: 1, category: "A" },
+      { id: 2, category: "B" },
+      { id: 3, category: "A" },
+    ];
+    const result = countBy(data, (x) => x.category);
+
+    expect(result).toStrictEqual({ A: 2, B: 1 });
+  });
+
+  test("mixed data types", () => {
+    const mySymbol = Symbol("mySymbol");
+    const data = [1, "a", 1, mySymbol, "A", mySymbol];
+    const result = countBy(data, (x) =>
+      typeof x === "string" ? x.toLowerCase() : x,
+    );
+
+    expect(result).toStrictEqual({ 1: 2, a: 2, [mySymbol]: 2 });
+  });
```

Comment:
Great test! But let's make it more concise by only testing a single concept each time. I think testing symbols is great, so I would like to have a test that only checks that aspect of the implementation.